File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    void fun(int n, List<String> ans, String s, int c, int o)
    {
        if(s.length() == 2*n)
        {
            ans.add(s);
            return;
        }
        if(o<n)
        {
            fun(n, ans, s+'(', c, o+1);
        }
        if(c<o)
        {
            fun(n, ans, s+')', c+1, o);
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> list = new ArrayList<>();
        fun(n, list, "", 0, 0);
        return list;
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []

        def fun(string, open, close, n):
            if len(string) == 2*n:
                ans.append(string)
                return
            
            if(open < n): fun(string+"(", open+1, close, n)
            if(open > close): fun(string+")", open, close+1, n)

        fun("", 0, 0, n)
        return ans